### PR TITLE
[command:depend-info] Add tree display

### PR DESCRIPTION
**Pull request by JackBoosY**
Example:
```pwsh
PS F:\vcpkg> .\vcpkg depend-info gdal --sort=x-tree
gdal
+-- cfitsio
|   +-- zlib
+-- curl
|   +-- zlib
+-- expat
+-- geos
+-- giflib
+-- hdf5
|   +-- szip
|   +-- zlib
+-- libgeotiff
|   +-- libjpeg-turbo
|   +-- proj4
|   |   +-- sqlite3
|   |   +-- sqlite3
|   |   +-- tiff
|   |       +-- libjpeg-turbo
|   |       +-- liblzma
|   |       +-- zlib
|   +-- tiff
|   |   +-- libjpeg-turbo
|   |   +-- liblzma
|   |   +-- zlib
|   +-- zlib
+-- liblzma
+-- libpng
|   +-- zlib
+-- libpq
|   +-- openssl
|   +-- zlib
+-- libwebp
+-- libxml2
|   +-- libiconv
|   +-- liblzma
|   +-- zlib
+-- netcdf-c
|   +-- curl
|   |   +-- zlib
|   +-- hdf5
|       +-- szip
|       +-- zlib
+-- openjpeg
+-- proj4
|   +-- sqlite3
|   +-- sqlite3
|   +-- tiff
|       +-- libjpeg-turbo
|       +-- liblzma
|       +-- zlib
+-- sqlite3
+-- zlib
+-- zstd
```

Current:
```
PS F:\vcpkg> .\vcpkg depend-info gdal
libjpeg-turbo:
liblzma:
zlib:
sqlite3[tool]:
szip:
tiff: libjpeg-turbo, liblzma, zlib
curl[non-http, ssl, schannel, sspi, winssl]: zlib
hdf5[szip, zlib]: szip, zlib
libiconv:
openssl:
proj4[database, tiff]: sqlite3, sqlite3, tiff
cfitsio: zlib
expat:
geos:
giflib:
libgeotiff: libjpeg-turbo, proj4, tiff, zlib
libpng: zlib
libpq[openssl, zlib]: openssl, zlib
libwebp[nearlossless, unicode, simd]:
libxml2: libiconv, liblzma, zlib
netcdf-c: curl, hdf5
openjpeg:
zstd:
gdal: cfitsio, curl, expat, geos, giflib, hdf5, libgeotiff, liblzma, libpng, libpq, libwebp, libxml2, netcdf-c, openjpeg, proj4, sqlite3, zlib, zstd
```

Fixes https://github.com/microsoft/vcpkg/issues/17678

### CODE REVIEW

**Comment by JackBoosY** [comment]
Do you guys have any suggestions?

File: src/vcpkg/commands.dependinfo.cpp [hunk 1]
```diff
@@ -304,36 +299,84 @@ namespace vcpkg::Commands::DependInfo
...
                 }
-                System::print2(": ", dependencies, "\n");
             }
         }
         Checks::exit_success(VCPKG_LINE_INFO);
     }
 
+    void RecurseFindDependencies(int level, std::string currDepend, std::vector<PackageDependInfo> allDepends)
+    {
+        auto currPos = allDepends.begin();
```
Comment (ras0219) [inline src/vcpkg/commands.dependinfo.cpp:358]:
```suggestion
        if (level == 100)
        {
            Checks::exit_with_message(VCPKG_LINE_INFO, "Recursion depth exceeded.");
        }
```

File: src/vcpkg/commands.dependinfo.cpp [hunk 2]
```diff
@@ -304,36 +299,84 @@ namespace vcpkg::Commands::DependInfo
...
+                    }
+                    System::print2(": ", dependencies, "\n");
                 }
-                System::print2(": ", dependencies, "\n");
             }
         }
         Checks::exit_success(VCPKG_LINE_INFO);
     }
 
+    void RecurseFindDependencies(int level, std::string currDepend, std::vector<PackageDependInfo> allDepends)
```
Comment (ras0219) [inline src/vcpkg/commands.dependinfo.cpp:356]:
```suggestion
    void RecurseFindDependencies(int level, const std::string& currDepend, const std::vector<PackageDependInfo>& allDepends)
```

File: src/vcpkg/commands.dependinfo.cpp [hunk 3]
```diff
@@ -304,36 +299,84 @@ namespace vcpkg::Commands::DependInfo
...
+    {
+        auto currPos = allDepends.begin();
+        for (currPos = allDepends.begin(); currPos != allDepends.end(); currPos++)
+        {
+            if (currPos->package == currDepend)
+            {
+                break;
+            }
+        }
```
Comment (ras0219) [inline src/vcpkg/commands.dependinfo.cpp:365]:
```suggestion
        auto currPos = std::find_if(allDepends.begin(), allDepends.end(), [&currDepend](const auto& p) { return p.package == currDepend; });
        Checks::check_exit(VCPKG_LINE_INFO, currPos != allDepends.end(), "internal vcpkg error");
```

File: src/vcpkg/commands.dependinfo.cpp [hunk 4]
```diff
@@ -304,34 +299,90 @@ namespace vcpkg::Commands::DependInfo
...
+                    System::print2("|   ");
+                else
+                    System::print2("    ");
             }
+
+            // Handle the current level
+            System::print2("+-- ", i->c_str(), "\n");
+            // The last element doesn't need additional connecting lines
+            auto j = i;
+            RecurseFindDependencies(levelFrom + 1, i->c_str(), ++j == currPos->dependencies.end(), allDepends);
```
Comment (JackBoosY) [inline src/vcpkg/commands.dependinfo.cpp:384]: `++j == currPos->dependencies.end()` maybe dangerous, but I don't have other solution.

File: src/vcpkg/commands.dependinfo.cpp [hunk 5]
```diff
@@ -304,34 +299,90 @@ namespace vcpkg::Commands::DependInfo
...
-                    System::print2(System::Color::warning, features);
-                    System::print2("]");
+                    System::print2("|   ");
                 }
-                System::print2(": ", dependencies, "\n");
+
+                if (!isEnd)
+                    System::print2("|   ");
+                else
+                    System::print2("    ");
```
Comment (JackBoosY) [inline src/vcpkg/commands.dependinfo.cpp:377]: Bug here:
```pwsh
+-- tesseract
|   +-- leptonica
|   |   +-- giflib
|   |   +-- libjpeg-turbo
|   |   +-- libpng
|   |   |   +-- zlib
|   |   +-- libwebp
|   |   +-- openjpeg
|   |   +-- tiff
|   |   |   +-- libjpeg-turbo
|   |   |   +-- liblzma
|   |   |   +-- zlib
|   |   +-- zlib
|   +-- libarchive
|       +-- bzip2
|       +-- liblzma
|       +-- libxml2
|   |   |   +-- libiconv
|   |   |   +-- liblzma
|   |   |   +-- zlib
|       +-- lz4
|   |   |   +-- xxhash
|       +-- openssl
|       +-- zlib
|       +-- zstd
```

**Review by ras0219-msft — APPROVED** [review]

File: src/vcpkg/commands.dependinfo.cpp (hunk 5 above)
Comment (ras0219-msft) [reply src/vcpkg/commands.dependinfo.cpp]: I fixed it :)

**Pull request merged**